### PR TITLE
Implement procedure for hashing arbitrary number of values

### PR DESCRIPTION
**Pull request by Overcastan**
This PR implements a new `hash_memory` procedure for `rpo` module which will calculate hash of the arbitrary number of values.

### CODE REVIEW

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 1]
```diff
@@ -100,3 +100,152 @@ export.hash_memory
   # drop start_addr/end_addr (4 cycles)
   movup.4 drop movup.4 drop
 end
+
+#! Computes hash of note inputs starting at the specified memory address.
```
Comment (Al-Kindi-0) [inline stdlib/asm/crypto/hashes/rpo.masm:104]: nit: note is maybe specific to miden-base so maybe we should change it

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 2]
```diff
@@ -100,3 +100,152 @@ export.hash_memory
...
+    
+    # prepare hasher state for RPO permutation 
+    push.0.0.0 padw padw 
+    # => [C, B, A, inputs_ptr, end_addr, num_inputs%8]
+
+    # hash every pair of words
+    exec.absorb_double_words_from_memory
+    # => [C', B', A', inputs_ptr', end_addr, num_inputs%8] where inputs_ptr' = end_addr
+
+    # hash remaining input values if there any left
```
Comment (Al-Kindi-0) [inline stdlib/asm/crypto/hashes/rpo.masm:135]: nit: ... if there are any left

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 3]
```diff
@@ -100,3 +100,152 @@ export.hash_memory
...
+export.hash_memory
+    # move number of inputs to the top of the stack 
+    swap
+    # => [num_inputs, inputs_ptr]
+
+    # get the number of double words
+    u32divmod.8 swap
+    # => [num_inputs/8, num_inputs%8, inputs_ptr]
+
+    # get the end_addr for hash_memory_even proc (end address for pairs of words)
```
Comment (Al-Kindi-0) [inline stdlib/asm/crypto/hashes/rpo.masm:119]: nit: ... for the hash_memory_even procedure

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 4]
```diff
@@ -100,3 +100,152 @@ export.hash_memory
...
+        dup.14 eq.1
+        # => [drop_counter == 1, latch = 0, e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # update the latch: if drop_counter == 1, latch will become 1
+        or
+        # => [latch', e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # save the latch value 
+        dup movdn.14
+        # => [latch', e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', latch', drop_counter]
```
Comment (Al-Kindi-0) [inline stdlib/asm/crypto/hashes/rpo.masm:186]: is there any reason as to why the latch needs to be this deep in the stack? why not on top of A'?

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 5]
```diff
@@ -100,3 +100,152 @@ export.hash_memory
...
+        # => [E, D, A', num_inputs%8]
+
+        # get the number of elements we need to drop
+        # notice that drop_counter could be any number from 1 to 7
+        push.8 movup.13 sub movdn.12
+        # => [E, D, A', drop_counter]
+
+        ### 0th value ########################################################
+        
+        # we need to drop first value anyway, since number of values is not divisible by 8
```
Comment (Al-Kindi-0) [inline stdlib/asm/crypto/hashes/rpo.masm:154]: could we add a remark about d3 i.e., if we are inside the if.true then d3 should be kept in any case and thus we can avoid treating in what follows (or something similar)

**Review by Al-Kindi-0 — CHANGES_REQUESTED** [review]
Very cool stuff, thank you!
Left minor comments and a couple of questions which I would like to understand better

File: stdlib/asm/crypto/hashes/rpo.masm (hunk 4 above)
Comment (Overcastan) [reply stdlib/asm/crypto/hashes/rpo.masm:186]: There is no actual reason, just intuitive desire not to divide RPO words with some other values. We can store it before A' (in position 9), but on both these cases we need to emulate operations (`movdn.9` and `movdn.14`). We can although store latch before `d_3`, in that case we will use `movdn.8` without any emulations, but not sure that this optimization worth it: in that case tracking of elements in the stack could become a little bit confusing for anyone who wants to understand how does it work. 

But I don't have any preferences, so maybe we indeed should store latch before `d_3`.

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 6]
```diff
@@ -100,3 +100,155 @@ export.hash_memory
...
+        # we have number of elements not divisible by 8, so we will have at least one element to 
+        # hash here (which turns out to be d_3)
+
+        hperm
+        # => [F, E, D]
+
+        exec.squeeze_digest
+        # => [E]
+    else
+        dropw movdnw.2 dropw dropw
```
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/rpo.masm:251]: Does this not drop an extra element? (based on line 133 there should be only 15 elements on the stack).

Also, can we incorporate `exec.squeeze_digest` into this (even if it takes an extra cycle).

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 7]
```diff
@@ -100,3 +100,155 @@ export.hash_memory
...
+        drop
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # push the padding 0 value 
+        push.0
+        # => [0, e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # move the calculated value down the stack 
+        movdn.6
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
```
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/rpo.masm:164]: The numbering of elements is opposite of what we usually do. Also, I'm not sure we benefit much from a more elaborate explanation here. I'd write this simply as:
```
drop push.0 movdn.6
# => [e_3, e_2, e_1, d_3, d_2, d_1, 0, d_0, A', drop_counter]
```

File: stdlib/tests/crypto/rpo.rs [hunk 8]
```diff
@@ -274,3 +274,30 @@ fn test_squeeze_digest() {
...
+
+        exec.rpo::hash_memory
+    end
+    ";
+
+    #[rustfmt::skip]
+    let expected_hash: Vec<u64> = build_expected_hash(&[
+        1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15
+    ]).into_iter().map(|e| e.as_int()).collect();
+    build_test!(compute_inputs_hash, &[]).expect_stack(&expected_hash);
```
Comment (bobbinth) [inline stdlib/tests/crypto/rpo.rs:302]: Could we also add a couple more test cases:

1. Hashing exactly 8 elements.
2. Hashing fewer than 8 elements (e.g., 6).

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few comments inline.

File: stdlib/tests/crypto/rpo.rs (hunk 8 above)
Comment (Overcastan) [reply stdlib/tests/crypto/rpo.rs:302]: Ah, my bad, I tested this proc on all variations from 0 to 16 elements, but forget to include it into the final test. I'll fix it.

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 9]
```diff
@@ -100,3 +100,161 @@ export.hash_memory
...
+    exec.absorb_double_words_from_memory
+    # => [C', B', A', ptr', end_addr, num_elements%8] where ptr' = end_addr
+
+    # hash remaining input values if there are any left
+    # if num_elements%8 is ZERO and there are no elements to hash
+    dup.14 eq.0
+    if.true
+      # clean the stack
+      exec.squeeze_digest
+      swapw dropw
```
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/rpo.masm:152]: I'm now sure I understand why `swapw dropw` works here. Isn't the state of the stack before these instructions look like so:
```
[B', ptr', end_addr, num_elements%8]
```
And so there are only 3 extra elements after `B'`. This means we drop one extra element from the stack - no?

Let's make sure we test for this in the tests.

Comment (Overcastan) [reply stdlib/asm/crypto/hashes/rpo.masm:152]: Ah, I see. I unconsciously assumed that there are no other elements on the stack, so everything except current values are `ZERO`'s so we can safely drop them. But in real life use cases there would be other elements, now I see what did you mean. I will fix this, sorry.

File: stdlib/tests/crypto/rpo.rs [hunk 10]
```diff
@@ -274,3 +274,87 @@ fn test_squeeze_digest() {
...
+
+        exec.rpo::hash_memory
+    end
+    ";
+
+    #[rustfmt::skip]
+    let expected_hash: Vec<u64> = build_expected_hash(&[
+        1, 2, 3, 4, 5
+    ]).into_iter().map(|e| e.as_int()).collect();
+    build_test!(compute_inputs_hash_5, &[]).expect_stack(&expected_hash);
```
Comment (bobbinth) [inline stdlib/tests/crypto/rpo.rs:298]: For this and other tests in this function, could we put an extra value onto the stack to make sure it is preserved. For example, this program could be:
```
use.std::crypto::hashes::rpo
begin
    push.1.2.3.4.1000 mem_storew dropw
    push.5.0.0.0.1001 mem_storew dropw
    push.11

    push.5.1000
    exec.rpo::hash_memory
end
```
And then the expected stack would be `[HASH, 11]`.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left one comment inline on how to make the tests a bit more robust. After that, we can merge.

**Pull request merged**